### ISSUE TITLE
SQL Admin reports "Failed to execute query" when database class contains indexed field

### ISSUE THREAD

**Issue by jedynykaban** [Starcounter/Home#456]
Starcounter version: `<2.3.2.90>`.

### Issue type

- [x ] Bug
- [ ] Feature request
- [ ] Suggestion
- [ ] Question
- [ ] Cannot reproduce
- [ ] Urgent

### Issue description
Problem is with SQL Admin, when in database class exists field with self-indexed field, SQL select query that conains this field cannot be executed (Failed to execute query / InvalidProgramException)

### Reproduction code snippet
Property code
```cs
public QuestionLang this[string langCode]
            => QuestionLangs?.FirstOrDefault(ql => ql.Language == langCode);
```
underneath is `Db.SQL<T>` call

### Exception stack trace

```
System.InvalidProgramException: Common Language Runtime detected an invalid program.
   at gen.Doctrin.Start.Server.Model.Questionnaires.Question.Doctrin.Start.Server.Model.Questionnaires.Question_Item_Binding.DoGetObject(Object )
   at Starcounter.Internal.DbState.View.GetObject(TypeBinding binding, Int32 index, Object proxy) in C:\TeamCity\BuildAgent\work\sc-13512\Level1\src\Starcounter\Internal\DbState.cs:line 71
   at Doctrin.Start.Server.Model.Questionnaires.Question.Starcounter.IObjectView.GetObject(Int32 index)
   at Starcounter.Query.Execution.Row.GetObject(Int32 index) in C:\TeamCity\BuildAgent\work\sc-13512\Level1\src\Starcounter\Query\Execution\Row\Row.cs:line 709
   at Starcounter.Internal.SqlRestHandler.ExecuteQuery(String query, Int32 maxResult) in C:\TeamCity\BuildAgent\work\sc-13512\Level1\src\Starcounter.Apps.JsonPatch\BuiltInRestHandlers\SqlRestHandler.cs:line 246
HResult=-2146233030
```

**Comment by k-rus** [comment]
Thank you for reporting the issue.
I don't know if this information will be enough to reproduce the issue. The application or schema with SQL query might be helpful.
@un-tone Can you look to this issue?

**Comment by jedynykaban** [comment]
I will try to provide You with more datails soon.

**Comment by un-tone** [comment]
@jedynykaban Please provide:
1. the SQL query you requested in SQL Admin
1. the index details
1. `QuestionLangs` query

**Comment by jedynykaban** [comment]
Hi sorry, I forgot a little about it. I prepared code (enclosed) to reproduce that bug, but You need to have at least one entity (data) in database, to made exception occur.

You can insert it with endpoint PUTting data like this:
`{
  "entities":[
    {
      "intValue": 5,
      "stringValue": "high five",
      "subId": 3,
      "stringSubValue": "low three"
    }
  ]
}`

Then just execute select from Admin:
`select x.* from ScAdminBugReproduce.Entity x`

In production `SubId` field is indexed in certain.
If You need more information just ping me please.

[ScAdminBugReproduce.zip](https://github.com/Starcounter/Home/files/2778440/ScAdminBugReproduce.zip)

**Comment by miyconst** [comment]
@jedynykaban thank you very much for the example for the issue reproduction. Unfortunately this is a limitation of current weaving technology.

Indexers in the database classes are not supported, the culprit code is this:

```cs
public virtual SubEntity this[long idx] => this
    .ExecuteRelationshipQuery<SubEntity>(nameof(SubEntity.Entity))
    .FirstOrDefault(x => x.SubId == idx);
```

You can replace it with a function like:

```cs
public virtual SubEntity GetSubEntity(long idx)
{
    return this
        .ExecuteRelationshipQuery<SubEntity>(nameof(SubEntity.Entity))
        .FirstOrDefault(x => x.SubId == idx);
}
```

cc @per-samuelsson

**Comment by per-samuelsson** [comment]
> Indexers in the database classes are not supported.

Don't see why we couldn't support those, because they are basically just syntax sugaring. Is it that weaver don't rule them out as database properties, or is it because SQL can't deal with them properly, in case we should expose them in the schema?

**Comment by miyconst** [comment]
@per-samuelsson, you are right, it actually might be Starcounter Administrator issue rather than weaver. I will check.

**Comment by jedynykaban** [comment]
Application works with enity defined this way. Like I reported, probelm starts, when I need to check data in SQLAdmin and hints:
* only occur when data are in tables (with empty tables Admin works ok),
* workaround is to declare interface explicitly (if there is such possiblity)), i.e. with small modifaction of attached code:
was:
`    [Database]
    public class Entity`
should be:
`    public interface IRelatedEntity<out T>
    {
        T this[long idx] { get; }
    }

    [Database]
    public class Entity : IRelatedEntity<SubEntity>`

You can change declaration of property from:
`public virtual SubEntity this[long idx] => this`
to:
`SubEntity IRelatedEntity<SubEntity>.this[long idx] => this`
and also Admin works fine (of course without presenting above field/property).
But! we cannot apply this solution, 'cause explicitly specified interface methods/properties cannot be "virtualized" (can be?), and then mocked for Unit Test,

**Comment by miyconst** [comment]
@jedynykaban thank you for the extra info, I confirm, that it's an issue in Starcounter Administrator.

**Comment by miyconst** [comment]
The issue is not applicable for Starcounter 3.0 and won't be fixed in 2.3.
